### ISSUE TITLE
Add random start to animatedTexture, or improve tileset animation

### ISSUE THREAD

**Issue by LeeWannacott** [godotengine/godot-proposals#8329]
### Describe the project you are working on

High Resolution style Top down 2d game... 

Basically, I want to animate water tiles with random start times at acceptable performance levels.

Using animated tileset/spritesheet gives ~120fps; However this option allows me to randomly start each tile which looks much nicer in my opinion. 

Using animatedTexture gives ~1200fps, but all tiles play at the same time.

 Note for animated tileset: options such as speed or Mode (default vs random start) doesn't seem to have a material effect); so, I assume its not the randomizing of tiles that is causing relatively bad performance.

**Animated texture ~1200fps  (but can't randomly start individual tiles)** 

https://github.com/godotengine/godot-proposals/assets/49783296/531cf254-cca7-4c40-9d5c-58ab3086a3db

**Animating from tileset (animated from a spritesheet) ~120fps; can randomize individual tiles start time :** 

https://github.com/godotengine/godot-proposals/assets/49783296/8571bcb6-1488-42d5-ba6c-f5f819c54ce5

I tried out two approaches to my problem:

**Animating tileset (animated from a spritesheet):** 
![image](https://github.com/godotengine/godot-proposals/assets/49783296/25c7428c-fb4d-46fa-aeb0-36fc38d200be)

**AnimatedTexture:**
![image](https://github.com/godotengine/godot-proposals/assets/49783296/dc02bce6-5a0f-497f-a978-0ab07935de4f)

Tested on: 5900x and 6800xt

### Describe the problem or limitation you are having in your project

I want to have ocean tiles that start their animation at random times (as it looks nicer) and also have acceptable performance, not just on my machine, but also on PC's less powerful...

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose either improving the performance of animated tiles from a tileset, or adding the option to randomize each individual tile with animated texture. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

For example: Add a similiar mode in texture dropdown as tileset animation has for random start.
![image](https://github.com/godotengine/godot-proposals/assets/49783296/0d56ebc8-fd87-46a7-b619-2562272e6fb9)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Possibly?, I welcome suggestions on how to implement this. Animated texture only gives you the current frame and randomizing current frame does it for all tiles, not individual tiles.

### Is there a reason why this should be core and not an add-on in the asset library?

Above my pay grade :shrug: . I don't think it would make sense as an add-on...

**Comment by Calinou** [comment]
cc @rakkarage

**Comment by LeeWannacott** [comment]
> The issue with AnimatedTexture is that it's a _resource_ that is shared across all drawn tiles, so randomizing its start offset wouldn't help in your particular case unless you create a few variants (4-8) and place them randomly in your TileMap as different tiles.

I did think of trying this it sounds doable, However not really ideal when you want to have maybe 4 variants for oceans which would each need 4+ varied rate AnimatedTexture; I think it could quickly get out of hand... and I'm not sure how it would look. I could try it out though... :thinking:

**Comment by dalexeev** [comment]
Also note that `AnimatedTexture` is marked as deprecated.

![](https://github.com/godotengine/godot-proposals/assets/47700418/40c298c2-5122-4cb4-8e00-fe4f62907040)
![](https://github.com/godotengine/godot-proposals/assets/47700418/9d0847b7-4e7d-408e-80b7-ed00d57ec27a)

**Comment by LeeWannacott** [comment]
`Warning: The current implementation is not efficient for the modern renderers.`

> Also note that `AnimatedTexture` is marked as deprecated.

Do you know what they are planning to replace it with, or any alternatives?

If it's TileSet animation then it might be good to have the one shot feature of the animatedTexture, for example weather effects; lets say you have an animation for snow falling that turns each corner of the tile into snow until it becomes a snow tile etc. You could also achieve this by having sub tiles within a tile, but Id imagine it would complicated to implement...

**Comment by Calinou** [comment]
> I thought VRAM compressing the images might help, but nope : /

The bottleneck isn't memory bandwidth, it's having to constantly upload new textures to the GPU continuously (which implies frequent CPU-GPU roundtrips).

**Comment by LeeWannacott** [comment]
> The bottleneck isn't memory bandwidth, it's having to constantly upload new textures to the GPU continuously (which implies frequent CPU-GPU roundtrips). 

Yeah, I thought VRAM compression would mean your images are in VRAM so the GPU can access them directly without hitting ram and CPU. Clearly, my understanding is wrong... 

> It's strongly suggested to do animation using shaders that modify UV over time instead.

I was trying to avoid writing a shader, it might be simple for cycling through some images though :shrug;, or do you mean having dynamic water using math? I tried to get animated water working from various shaders on godot shaders, I couldn't get many working and the ones that I could either didn't look appropriate, or/and crippled performance especially those designed for 3d ones.

slightly off topic: I feel like I'm hitting so many issues I have to work around... would you say that godot is appropriate for an ambitous 2D game in the vein of say dwarf fortress, factorio,  rimworld, songs of syx, or is it meant for smaller games, or just needs more dev time?

**Comment by Calinou** [comment]
> I was trying to avoid writing a shader, it might be simple for cycling through some images though :shrug;, or do you mean having dynamic water using math?

Both approaches are viable, though it seems you want to go for premade images if you want more artistic control.

> slightly off topic: I feel like I'm hitting so many issues I have to work around... would you say that godot is appropriate for an ambitous 2D game in the vein of say dwarf fortress, factorio, rimworld, songs of syx, or is it meant for smaller games, or just needs more dev time?

Every complex game has engine issues to work around, no matter which engine you use. Gamedev is a game of workarounds :slightly_smiling_face:
